### PR TITLE
[analyzer][NFC] Migrate nonloc::ConcreteInt to use APSIntPtr (2/4)

### PR DESCRIPTION
**Pull request by steakhal**

### CODE REVIEW

**Comment by steakhal** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/120436?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#120438** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/120438?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#120437** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/120437?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#120436** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/120436?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/120436?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#120435** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/120435?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h [hunk 1]
```diff
@@ -298,9 +299,12 @@ class SymbolVal : public NonLoc {
 /// Value representing integer constant.
 class ConcreteInt : public NonLoc {
 public:
-  explicit ConcreteInt(const llvm::APSInt &V) : NonLoc(ConcreteIntKind, &V) {}
+  explicit ConcreteInt(APSIntPtr V) : NonLoc(ConcreteIntKind, V.get()) {}
 
-  const llvm::APSInt &getValue() const { return *castDataAs<llvm::APSInt>(); }
+  APSIntPtr getValue() const {
+    // This is safe because in the ctor we take a safe APSIntPtr.
+    return APSIntPtr::unsafeConstructor(castDataAs<llvm::APSInt>());
```
Comment (Xazax-hun) [inline clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:306]: Is it possible to store `ApsIntPtr` in `NonLoc`?

Comment (Xazax-hun) [reply clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:306]: > Currently the infrastructure is set up for holding raw-pointers.

Could you elaborate on what preventing us from changing this?

Comment (steakhal) [reply clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:306]: Basicall it all boils down to the fact that an sval has a single void pointer as a storage. I'm currently okay casting a pointer to and from void pointer of this storage. But I'd be uncomfortable bitcasting an APSIntPtr object into a void pointer storage, even if it only holds a single pointer inside. This is why I preferred unboxing the pointer and holding that inside the SVal.

Maybe I overlooked some other techniques. Let me know if you have something in mind.

Comment (steakhal) [reply clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:306]: A union make this void pointer trickery marginally safer at the expense of declaring all the types we typepun this poor void storage pointer.
Maybe this is something to consider.

Comment (Xazax-hun) [reply clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:306]: According to [CppReference](https://en.cppreference.com/w/c/language/struct):
> A pointer to a struct can be cast to a pointer to its first member (or, if the member is a bit-field, to its allocation unit). Likewise, a pointer to the first member of a struct can be cast to a pointer to the enclosing struct.

Comment (steakhal) [reply clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:306]: I see now. That was the source of the miscommunication here.

**Review by Xazax-hun — APPROVED** [review]

File: clang/lib/StaticAnalyzer/Checkers/ArrayBoundCheckerV2.cpp [hunk 2]
```diff
@@ -241,26 +242,25 @@ computeOffset(ProgramStateRef State, SValBuilder &SVB, SVal Location) {
...
-              svalBuilder.makeIntVal(extent.getValue() / constant),
-              svalBuilder);
+              svalBuilder.makeIntVal(extentVal / constant), svalBuilder);
       case BO_Add:
         return getSimplifiedOffsets(
             nonloc::SymbolVal(SIE->getLHS()),
-            svalBuilder.makeIntVal(extent.getValue() - constant), svalBuilder);
+            svalBuilder.makeIntVal(extentVal - constant), svalBuilder);
       default:
         break;
```
Comment (NagyDonat) [inline clang/lib/StaticAnalyzer/Checkers/ArrayBoundCheckerV2.cpp:265]: Slightly offtopic request: as you're already editing many lines of this function, please convert it to `UpperCamelCase` variable names (which is the standard suggested by the LLVM style guide).

When I started to work on this checker in April 2023, it was an ugly mixture of `UpperCamelCase` and `lowerCamelCase` variable names; but then, as I refactored various parts of it, I gradually renamed (or removed) most old `lowerCamelCase` variables, so (if I recall correctly) this is the last function that heavily uses them.

**Comment by steakhal** [comment]
### Merge activity

* **Dec 19, 6:49 AM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/120436).

**Pull request merged**